### PR TITLE
Federation registry role

### PR DESCRIPTION
**Pull request by linathedog**
Added federation-demo role that deploys the latest version of the federation-registry project. In more details the role performs the following actions:
1) It prepares the target host by installing all required packages and tools
2) It clones the project from github
3) It installs and configures the node projects (federation-registry-frontend,federation-registry-backend-api,federation-registry-backend-ams-agent)
4) It creates and configures the postgresql database if necessary
5) It creates and configures the ams topics and subscriptions if necessary

### CODE REVIEW

File: roles/federation-demo/defaults/main.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+db_username: postgres
+db_password: postgres
+db_host: localhost
+ams_agent_key : "authrization_key"
+ams_token : "ams_token"
+ams_host: "https://msg-devel.argo.grnet.gr/v1"
+ams_project: "rciam-service-registry"
+ams_sub: "express_push"
+ams_topic: "deployment_response"
+db_script_path: /home/debian/setup_db.sql
```
Comment (Tas-sos) [inline roles/federation-demo/defaults/main.yml:22]: If you want, it suits to your use cases and you agree that it is more beautiful, I would suggest you to "group" the variables in lists, for example:
```
git :
  REPO: "git@github.com:rciam/rciam-federation-registry.git"
  BRANCH: "devel"
  GITHUB_ACCESS_TOKEN: "token"
  CLONE_DEST: "/var/www/rciam-federation-registry"
  GIT_EXECUTABLE: "/usr/bin/git"

db:
  name: service_registry_db
  username: postgres
  password: postgres
  host: localhost
  script_path: /home/debian/setup_db.sql

ams:
  agent_key : "authrization_key"
  token : "ams_token"
  host: "https://msg-devel.argo.grnet.gr/v1"
  project: "rciam-service-registry"
  sub: "express_push"
  topic: "deployment_response"

KEY_TITLE: "evironment-vm"
KEY_PATH: "~/.ssh/id_rsa.git"
deployment_env : "test"

KNOWN_HOSTS_PATH: "~/.ssh/known_hosts"
```

**Review by Tas-sos — COMMENTED** [review]
Bravo @linathedog! Very :muscle: work! :clap:

File: roles/federation-demo/tasks/deploy.yml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    register: project_stat
+
+  - name: Clean var/www
+    file:
+      state: absent
+      path: "{{ CLONE_DEST }}"
+    when: project_stat.stat.exists
+
+  - name: Clone the repository
+    shell: GIT_SSH_COMMAND="ssh -i {{ KEY_PATH }} -v -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null" {{ GIT_EXECUTABLE }} clone --single-branch --branch {{GIT_BRANCH}} {{ GIT_REPO }} {{ CLONE_DEST }}
```
Comment (Tas-sos) [inline roles/federation-demo/tasks/deploy.yml:14]: I would suggest you to use the **git** Ansible mode :
https://docs.ansible.com/ansible/2.9/modules/git_module.html

**Comment by NicolasLiampotis** [comment]
@linathedog would it be possible to have a single role for deploying both the demo and the production environment of the federation registry? For example:

Production vars (should be the defaults though `defaults/main.yml`):
```
federation_git:
  git_repo: "git@github.com/rciam/rciam-federation-registry.git"
  git_branch: "master"
```
Demo vars:
```
federation_git:
  git_repo: "git@github.com/rciam/rciam-federation-registry.git"
  git_branch: "devel"
```

File: roles/federation-demo/tasks/deploy.yml [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      state: absent
+      path: "{{ federation_git.clone_dest }}"
+    when: project_stat.stat.exists
+
+  - name: Clone the repository
+    git:
+      repo: "{{ federation_git.git_repo }}"
+      version: "{{federation_git.git_branch}}"
+      dest: "{{federation_git.clone_dest}}"
+      key_file: "{{ federation_git.key_path }}"
```
Comment (Tas-sos) [inline roles/federation-demo/tasks/deploy.yml:18]:
```suggestion
  - name: Check if project exists
    stat:
      path: "{{ federation_git.clone_dest }}"
    register: project_stat

  - name: Clean clone directory
    file:
      state: absent
      path: "{{ federation_git.clone_dest }}"
    when: project_stat.stat.exists

  - name: Clone the repository
    git:
      repo: "{{ federation_git.git_repo }}"
      version: "{{federation_git.git_branch}}"
      dest: "{{federation_git.clone_dest}}"
      key_file: "{{ federation_git.key_path }}"
      force: yes
```

If you want, you can check the [force](https://docs.ansible.com/ansible/2.9/modules/git_module.html#parameter-force)`:yes` parameter.

File: roles/federation-demo/tasks/configure-environment.yml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    apt:
+      pkg:
+        - postgresql
+        - postgresql-contrib
+        - curl
+        - git
+        - nginx
+        - python-pip
+
+
```
Comment (Tas-sos) [inline roles/federation-demo/tasks/configure-environment.yml:12]:
```suggestion
```

File: roles/federation-demo/tasks/deploy.yml [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      path: "{{federation_express_path}}"
+
+  - name: Install node modules for ams-agent
+    npm:
+      path: "{{federation_ams_agent_path}}"
+
+  - name: React config
+    template:
+      src: react-config.j2
+      dest: "{{federation_react_path}}/src/config.json"
```
Comment (Tas-sos) [inline roles/federation-demo/tasks/deploy.yml:36]:
```suggestion
      dest: "{{federation_react_path}}/src/config.json"
      backup: yes
```

I suggest you use the `backup` parameter on most of the time ( template/file/copy/lineinfile )
https://docs.ansible.com/ansible/2.9/modules/template_module.html#parameter-backup

File: roles/federation-demo/tasks/deploy.yml [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+  - name: Configure express db_config
+    template:
+      src: db_config.j2
+      dest: "{{federation_express_path}}/db-config/db-config.json"
```
Comment (Tas-sos) [inline roles/federation-demo/tasks/deploy.yml:41]:
```suggestion
      dest: "{{federation_express_path}}/db-config/db-config.json"
      backup: yes
```

File: roles/federation-demo/tasks/deploy.yml [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+  - name: Expres env
+    template:
+      src: express-env.j2
+      dest: "{{federation_express_path}}/JavaScript/.env"
```
Comment (Tas-sos) [inline roles/federation-demo/tasks/deploy.yml:46]:
```suggestion
      dest: "{{federation_express_path}}/JavaScript/.env"
      backup: yes
```

File: roles/federation-demo/tasks/deploy.yml [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+  - name: Agent env
+    template:
+      src: ams-env.j2
+      dest: "{{federation_ams_agent_path}}/.env"
```
Comment (Tas-sos) [inline roles/federation-demo/tasks/deploy.yml:51]:
```suggestion
      dest: "{{federation_ams_agent_path}}/.env"
      backup: yes
```

File: roles/federation-demo/tasks/deploy.yml [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+  - name: delete existing pm2 processes if running
+    command: "pm2 delete all"
+    ignore_errors: True
+
+  - name: Create pm2 ecosystem
+    template:
+     src: pm2_config.j2
+     dest: "{{federation_pm2_conf_path}}/ecosystem.config.js"
```
Comment (Tas-sos) [inline roles/federation-demo/tasks/deploy.yml:60]:
```suggestion
     dest: "{{federation_pm2_conf_path}}/ecosystem.config.js"
     backup: yes
```

File: roles/federation-demo/tasks/postgres.yml [hunk 10]
```diff
@@ -0,0 +1,47 @@
+---
+- name: Make sure psycopg2 is installed
+  command: pip install psycopg2==2.7.5 --ignore-installed
+
+- name: Send file
+  copy:
+    src: private_files/setup_db.sql
+    dest: "{{federation_db.db_script_path}}"
```
Comment (Tas-sos) [inline roles/federation-demo/tasks/postgres.yml:8]:
```suggestion
    dest: "{{federation_db.db_script_path}}"
    backup: yes
```

https://docs.ansible.com/ansible/2.9/modules/copy_module.html#parameter-backup

File: roles/federation-demo/tasks/postgres.yml [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    db: "{{ federation_db.db_name }}"
+    login_user: "{{ federation_db.db_username }}"
+    login_password: "{{ federation_db.db_password }}"
+    path_to_script: "{{ federation_db.db_script_path }}"
+    login_host: "{{ federation_db.db_host }}"
+  register: postgresql_result
+  failed_when: "postgresql_result is failed and 'already exists' not in postgresql_result.msg"
+  when: "federation_db.db_name is defined"
+
+- name: Send file
```
Comment (Tas-sos) [inline roles/federation-demo/tasks/postgres.yml:28]:
```suggestion
- name: Send a magic file ^_^
```

Could you have a more representative name/message here please ?

File: roles/federation-demo/tasks/postgres.yml [hunk 12]
```diff
@@ -0,0 +1,47 @@
+---
+- name: Make sure psycopg2 is installed
+  command: pip install psycopg2==2.7.5 --ignore-installed
```
Comment (Tas-sos) [inline roles/federation-demo/tasks/postgres.yml:3]: You can also use the [pip ansible module](https://docs.ansible.com/ansible/2.9/modules/pip_module.html).

----------------------
References :
 
https://github.com/rciam/rciam-deploy/blob/a4e475a63833748108288f6b1fdc41b918001003/roles/fedreg-agent/tasks/deploy.yml#L16-L38

https://github.com/rciam/rciam-deploy/blob/a4e475a63833748108288f6b1fdc41b918001003/roles/fedreg-agent/tasks/configure-common.yml#L18-L38

File: roles/federation-demo/tasks/deploy.yml [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+  - name: start Pm2 processes
+    command: "pm2 start ecosystem.config.js"
+
+  - name: restart Pm2 processes
+    command: "pm2 restart all"
+
+  - name: react build
+    command: "npm --prefix {{federation_react_path}} run build"
```
Comment (Tas-sos) [inline roles/federation-demo/tasks/deploy.yml:69]: I suggest you the following:
```suggestion
    shell: "npm run build"
      args:
        chdir: "{{ federation_react_path }}"
        executable: /bin/bash
```

* https://docs.ansible.com/ansible/2.9/modules/shell_module.html#parameter-chdir
* https://docs.ansible.com/ansible/2.9/modules/command_module.html#parameter-chdir

**Review by NicolasLiampotis — CHANGES_REQUESTED** [review]
Would it be possible to have a single role for the federation-registry and use different inventories/vars to handle the different environments (devel, demo, production)?

**Review by NicolasLiampotis — CHANGES_REQUESTED** [review]
1. Rename role to federation-registry
2. Use prefix federation_registry_ prefix for all variables. E.g. `federation_git.git_branch` becomes `federation_registry_git.branch` (or perhaps `federation_registry_git.version` would be more appropriate for this specific var)

File: roles/federation-registry/defaults/main.yml [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+  github_access_token: "{{vault_GITHUB_ACCESS_TOKEN}}"
+  key_path: "~/.ssh/id_rsa.git"
+  git_repo: "git@github.com:linathedog/rciam-federation-registry.git"
+  git_branch: "devel"
+  clone_dest: "/var/www/rciam-federation-registry"
+federation_registry_known_hosts_path: "~/.ssh/known_hosts"
+federation_registry_db:
+  name: service_registry_db
+  username: postgres
+  password: postgres
```
Comment (Tas-sos) [inline roles/federation-registry/defaults/main.yml:14]: I suggest you do not use easy usernames and passwords.
I also suggest encrypting these variables with the Ansible vault.

File: roles/federation-registry/tasks/postgres.yml [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+  copy:
+    src: private_files/setup_db.sql
+    dest: "{{federation_registry_db.script_path}}"
+
+- name: Create a new database with name service_registry_db
+  postgresql_db:
+    name: service_registry_db
+  become: yes
+  become_user: postgres
+
```
Comment (NicolasLiampotis) [inline roles/federation-registry/tasks/postgres.yml:15]: Remove. DB creation will be handled externally (e.g. with rciam/postgres role or the db is provided externally

File: roles/federation-registry/defaults/main.yml [hunk 16]
```diff
@@ -0,0 +1,30 @@
+---
+# defaults file for federation-demo
+federation_registry_key_title: "evironment-vm"
+federation_registry_git:
+  github_access_token: "{{vault_GITHUB_ACCESS_TOKEN}}"
```
Comment (NicolasLiampotis) [inline roles/federation-registry/defaults/main.yml:5]: change "{{vault_GITHUB_ACCESS_TOKEN}}"  to "secret_access_token"

**Review by NicolasLiampotis — COMMENTED** [review]
Add comments in defaults to document config options

**Review by NicolasLiampotis — APPROVED** [review]

**Pull request merged**